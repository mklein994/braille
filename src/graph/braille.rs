//! Turn a series of numbers into a braille graph.
//!
//! # Example
//!
//! ## Sine graph
//!
//! ```console
//! $ awk 'BEGIN { for (i = 0; i < 20; i++) { print sin(i / 3); } }' | braille -1 1 4
//! ⠀⠀⠀⠀⠀⣷⣶⣤⣀
//! ⠀⠀⠀⠀⠀⣿⣿⣿⡿⠟
//! ⠀⠀⢀⣀⣤⡟⠉⠁
//! ⣴⣿⣿⣿⣿⡇
//! ⠀⠉⠛⠻⠿⡇
//! ```
//!
//! ## Simple number sequence
//!
//! ```console
//! $ seq -4 3 | braille -4 3 7
//! ⠉⠛⠿⣿
//! ⠀⠀⠀⢸⣶⣤⣀
//! ```

use super::{ColumnGraphable, Graphable};
use crate::opt::Config;
use crate::LineResult;

#[derive(Debug)]
pub struct Lines {
    config: Config,
}

impl Graphable for Lines {
    type Config = Config;
    type Item = f64;

    fn new(config: Self::Config) -> Self {
        Self { config }
    }

    fn config(&self) -> &Self::Config {
        &self.config
    }
}

impl ColumnGraphable for Lines {
    fn print_lines(&self, mut input_lines: impl Iterator<Item = LineResult>) -> anyhow::Result<()> {
        let minimum = self.minimum();
        let maximum = self.maximum();

        let min = 1; // reserve an empty line for null values
        let max = self.width() * 2; // braille characters are 2 dots wide
        let slope = f64::from(max - min) / (maximum - minimum);
        let scale = |value: f64| {
            assert!(
                value >= minimum && value <= maximum,
                "value out of bounds: {value} [{minimum}, {maximum}]"
            );
            min + (slope * (value - minimum)).round() as u16
        };

        // Clamp where 0 would fit to be inside the output range
        let zero = if minimum > 0. {
            min
        } else if maximum < 0. {
            max
        } else {
            scale(0.)
        };

        // Each braille character is 4 dots tall
        let mut buffer = [vec![], vec![], vec![], vec![]];
        let mut has_more_lines = true;
        while has_more_lines {
            for buffer_line in &mut buffer {
                let input_line = input_lines.next();
                if input_line.is_none() {
                    has_more_lines = false;
                }

                if let Some(new_line) = input_line
                    .transpose()?
                    .flatten()
                    .map(scale)
                    .map(|value| Self::into_dot_pairs(value, zero))
                {
                    *buffer_line = new_line;
                }
            }

            if has_more_lines || buffer.iter().any(|x| !x.is_empty()) {
                let transposed = Self::transpose_row(&buffer);
                let braille_char = Self::to_braille_char_row(&transposed);
                println!("{braille_char}");
            }

            buffer.fill(vec![]);
        }

        Ok(())
    }
}

impl Lines {
    /// Turn a value into its representation of braille dots for that row
    ///
    /// # Example:
    ///
    /// Let's say the width is 4 characters wide, and the input is all integers in the range `[-3, 4]`
    /// incrementing by one. The braille pattern would look like this:
    ///
    /// ```plain
    /// ⠙⢿
    /// ⠀⢸⣷⣄
    /// ```
    ///
    /// Depending on your font, this may appear with gaps between lines or characters. In my
    /// experience, [Cascadia Code](https://github.com/microsoft/cascadia-code) looks really good.
    ///
    /// Breaking this down, let's look at these lines in particular:
    ///
    /// | Value | ASCII braille | Notes            |
    /// |------:|:--------------|------------------|
    /// |    -3 | `** **`       |                  |
    /// |    -2 | `-* **`       | 1st example row  |
    /// |    -1 | `-- **`       |                  |
    /// |     0 | `-- -*`       |                  |
    /// |       |               | _next character_ |
    /// |     1 | `-- -* *- --` |                  |
    /// |     2 | `-- -* ** --` |                  |
    /// |     3 | `-- -* ** *-` | 2nd example row  |
    /// |     4 | `-- -* ** **` |                  |
    ///
    /// In the first example row, the value -2 translates to dot 4. This means dots 2 through 4 are
    /// filled, and any before that are blank. Since the value is below zero, it looks like this:
    ///
    /// ```
    /// //   ┌──── -2 (dot 2)
    /// //  -* **
    /// //      └─  0 (dot 4)
    /// assert_eq!(vec![[false, true], [true, true]], braille::Lines::into_dot_pairs(2, 4));
    /// ```
    ///
    /// In the second example row, the input value is 3, which translates to dot 7. The value is above
    /// zero, and looks like this:
    ///
    /// ```
    /// //          ┌── 3 (dot 7)
    /// // -- -* ** *-
    /// //     └─────── 0 (dot 4)
    /// assert_eq!(
    ///     vec![[false, false], [false, true], [true, true], [true, false]],
    ///     braille::Lines::into_dot_pairs(7, 4)
    /// );
    /// ```
    ///
    /// Looking at each example line more closely, we can see it broken down into parts:
    ///
    /// ```plain
    ///     ┌───── prefix
    /// 1.  -* **
    ///      └─┴┴─ stem
    ///
    ///     ┌┬─┬──────── prefix
    ///     ││ │     ┌┬─ tip
    /// 2.  -- -* ** *-
    ///         └─┴┴──── stem
    /// ```
    #[must_use]
    pub fn into_dot_pairs(value: u16, zero: u16) -> Vec<[bool; 2]> {
        let prefix_length = usize::from(value.min(zero) - 1);
        let mut iter = vec![false; prefix_length];

        let stem_length = usize::from(value.abs_diff(zero) + 1);
        iter.resize(iter.len() + stem_length, true);

        let chunks = iter.chunks_exact(2);
        let tip = chunks.remainder().to_vec();
        let mut row: Vec<[bool; 2]> = chunks
            .into_iter()
            .map(|chunk| [chunk[0], chunk[1]])
            .collect();
        if !tip.is_empty() {
            row.push([tip[0], tip.get(1).copied().unwrap_or_default()]);
        }
        row
    }

    /// Turns an array of dot pairs into a braille character.
    ///
    /// # Example
    ///
    /// ```
    /// assert_eq!(
    ///     braille::Lines::to_braille_char([
    ///         [true, true],
    ///         [false, true],
    ///         [true, false],
    ///         [true, true],
    ///     ]),
    ///     '⣝'
    /// );
    /// ```
    ///
    /// See also: <https://en.wikipedia.org/wiki/Braille_Patterns>
    #[must_use]
    pub fn to_braille_char(dot_pairs: [[bool; 2]; 4]) -> char {
        // Turn this:
        //
        // ```plain
        // [
        //   [0, 3],
        //   [1, 4],
        //   [2, 5],
        //   [6, 7],
        // ]
        // ```
        //
        // into this:
        //
        // ```plain
        // [0, 1, 2, 3, 4, 5, 6, 7]
        // ```
        let bits = [
            dot_pairs[0][0],
            dot_pairs[1][0],
            dot_pairs[2][0],
            dot_pairs[0][1],
            dot_pairs[1][1],
            dot_pairs[2][1],
            dot_pairs[3][0],
            dot_pairs[3][1],
        ];

        let mut block = 0x2800_u32; // empty braille character

        for (index, bit) in bits.iter().enumerate() {
            if *bit {
                let position = u32::try_from(index).unwrap();
                block += (2_u32).pow(position);
            }
        }

        char::from_u32(block).expect("braille character not valid")
    }

    /// Render a list of braille dot blocks as a string
    #[must_use]
    pub fn to_braille_char_row(transposed: &[[[bool; 2]; 4]]) -> String {
        let mut line = String::new();
        for dot_pairs in transposed {
            let braille_char = Self::to_braille_char(*dot_pairs);

            line.push(braille_char);
        }
        line
    }

    /// Turn a list of braille dot pairs into a list of braille dot blocks
    #[must_use]
    pub fn transpose_row(input_row: &[Vec<[bool; 2]>; 4]) -> Vec<[[bool; 2]; 4]> {
        let longest = input_row.iter().map(Vec::len).max().unwrap();

        let mut output_row: Vec<[[bool; 2]; 4]> = vec![];
        for column in 0..longest {
            let mut braille_character = [[false, false]; 4];

            for (row_index, row) in input_row.iter().enumerate() {
                if let Some(row_column) = row.get(column) {
                    braille_character[row_index] = *row_column;
                }
            }

            if column < longest - 1 || braille_character.into_iter().flatten().any(|x| x) {
                output_row.push(braille_character);
            }
        }

        output_row
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_into_dot_pairs() {
        assert_eq!(
            vec![[false, false], [false, false], [true, false]],
            Lines::into_dot_pairs(5, 5)
        );
    }

    #[test]
    fn test_transpose_row_line_1() {
        #[rustfmt::skip]
        let input = [
            vec![[ true,  true], [ true, true], [true, false]],
            vec![[false,  true], [ true, true], [true, false]],
            vec![[false, false], [ true, true], [true, false]],
            vec![[false, false], [false, true], [true, false]],
        ];

        #[rustfmt::skip]
        let expected = vec![
            [
                [ true,  true],
                [false,  true],
                [false, false],
                [false, false],
            ],
            [
                [ true,  true],
                [ true,  true],
                [ true,  true],
                [false,  true],
            ],
            [
                [ true, false],
                [ true, false],
                [ true, false],
                [ true, false],
            ],
        ];
        let actual = Lines::transpose_row(&input);
        for (ex, act) in expected.iter().zip(actual.iter()) {
            eprintln!("{ex:?}");
            eprintln!("{act:?}");
            eprintln!();
        }
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_transpose_row_line_2() {
        #[rustfmt::skip]
        let input = [
            vec![[false, false], [false, false], [true, false]               ],
            vec![[false, false], [false, false], [true,  true]               ],
            vec![[false, false], [false, false], [true,  true], [true, false]],
            vec![[false, false], [false, false], [true,  true], [true,  true]],
        ];

        #[rustfmt::skip]
        let expected = vec![
            [
                [false, false],
                [false, false],
                [false, false],
                [false, false],
            ],
            [
                [false, false],
                [false, false],
                [false, false],
                [false, false],
            ],
            [
                [ true, false],
                [ true,  true],
                [ true,  true],
                [ true,  true]
            ],
            [
                [false, false],
                [false, false],
                [ true, false],
                [ true,  true],
            ],
        ];

        assert_eq!(expected, Lines::transpose_row(&input));
    }
}
