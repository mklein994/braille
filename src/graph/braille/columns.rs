use std::io::LineWriter;
use std::io::Write;

use crate::opt::ValueIter;
use crate::Config;
use crate::{ColumnGraphable, Graphable};

pub struct Columns {
    config: Config,
}

impl Graphable<Option<f64>> for Columns {
    fn new(config: Config) -> Self {
        Self { config }
    }

    fn config(&self) -> &Config {
        &self.config
    }

    fn print_graph(&self, lines: ValueIter<Option<f64>>) -> anyhow::Result<()> {
        let minimum = <Self as Graphable<Option<f64>, Config>>::minimum(self);
        let maximum = <Self as Graphable<Option<f64>, Config>>::maximum(self);

        let min = 1;
        let max = self.height() * 4;
        let slope = f64::from(max - min) / (maximum - minimum);
        let scale = |value: f64| {
            assert!(
                value >= minimum && value <= maximum,
                "value out of bounds: {value} [{minimum}, {maximum}]"
            );
            min + (slope * (value - minimum)).round() as u16
        };

        // Clamp where 0 would fit to be inside the output range
        let zero = if minimum > 0. {
            min
        } else if maximum < 0. {
            max
        } else {
            scale(0.)
        };

        let mut input_lines = lines.into_iter();

        let mut column_quads = vec![];

        loop {
            let left = input_lines.next();
            let right = input_lines.next();
            if right.is_none() {
                break;
            }

            let mut column = [vec![], vec![]];
            for (i, side) in [left, right].into_iter().enumerate() {
                if let Some(value) = side.transpose()?.and_then(|x| x.into_inner()).map(scale) {
                    column[i] = Self::into_dot_quads(value, zero);
                }
            }

            column_quads.push(column);
        }

        let mut writer = LineWriter::new(std::io::stdout());
        Self::into_braille_rows(&mut writer, &column_quads, usize::from(self.height()))?;

        Ok(())
    }
}

impl<const N: usize> Graphable<[Option<f64>; N]> for Columns {
    fn new(config: Config) -> Self {
        Self { config }
    }

    fn config(&self) -> &Config {
        &self.config
    }

    fn print_graph(&self, lines: ValueIter<[Option<f64>; N]>) -> anyhow::Result<()> {
        let minimum = <Self as Graphable<Option<f64>, Config>>::minimum(self);
        let maximum = <Self as Graphable<Option<f64>, Config>>::maximum(self);

        let min = 1;
        let max = self.height() * 4;
        let slope = f64::from(max - min) / (maximum - minimum);
        let scale = |value: f64| {
            assert!(
                value >= minimum && value <= maximum,
                "value out of bounds: {value} [{minimum}, {maximum}]"
            );
            min + (slope * (value - minimum)).round() as u16
        };

        let mut input_lines = lines.into_iter();

        let mut column_quads = vec![];

        loop {
            let left = input_lines.next();
            let right = input_lines.next();
            if right.is_none() {
                break;
            }

            let mut column = [vec![], vec![]];
            for (i, side) in [left, right].into_iter().enumerate() {
                if let Some(value) = side.transpose()?.map(|input_line_value| {
                    input_line_value
                        .into_iter()
                        .map(|x| x.map(scale).unwrap_or_default())
                        .collect::<Vec<_>>()
                        .try_into()
                        .unwrap()
                }) {
                    column[i] = Self::into_dot_quads_from_array::<N>(value);
                }
            }

            column_quads.push(column);
        }

        let mut writer = LineWriter::new(std::io::stdout());
        Self::into_braille_rows(&mut writer, &column_quads, usize::from(self.height()))?;

        Ok(())
    }
}

impl ColumnGraphable<Option<f64>> for Columns {}

impl Columns {
    fn into_dot_quads(value: u16, zero: u16) -> Vec<[bool; 4]> {
        let prefix_length = usize::from(value.min(zero) - 1);
        let mut iter = vec![false; prefix_length];

        let stem_length = usize::from(value.abs_diff(zero) + 1);
        iter.resize(iter.len() + stem_length, true);

        let chunks = iter.chunks_exact(4);
        let tip = chunks.remainder().to_vec();
        let mut column: Vec<[bool; 4]> = chunks
            .into_iter()
            .map(|chunk| [chunk[0], chunk[1], chunk[2], chunk[3]])
            .collect();
        if !tip.is_empty() {
            column.push([
                tip[0],
                tip.get(1).copied().unwrap_or_default(),
                tip.get(2).copied().unwrap_or_default(),
                tip.get(3).copied().unwrap_or_default(),
            ]);
        }

        column
    }

    fn into_braille_rows<W: ?Sized + Write>(
        line_writer: &mut LineWriter<W>,
        column_quads: &[[Vec<[bool; 4]>; 2]],
        height: usize,
    ) -> std::io::Result<()> {
        for row_index in (0..height).rev() {
            for col in column_quads {
                let mut raw_braille_char = [[false; 2]; 4];
                for (character_row, pair) in raw_braille_char.iter_mut().rev().enumerate() {
                    let left = col
                        .first()
                        .and_then(|c| c.get(row_index))
                        .and_then(|c| c.get(character_row))
                        .copied()
                        .unwrap_or_default();
                    let right = col
                        .last()
                        .and_then(|c| c.get(row_index))
                        .and_then(|c| c.get(character_row))
                        .copied()
                        .unwrap_or_default();
                    *pair = [left, right];
                }

                write!(
                    line_writer,
                    "{}",
                    super::Lines::to_braille_char(raw_braille_char)
                )?;
            }

            writeln!(line_writer)?;
        }

        line_writer.flush()?;

        Ok(())
    }

    fn into_dot_quads_from_array<const N: usize>(line_set: [u16; N]) -> Vec<[bool; 4]> {
        assert_eq!(2, line_set.len(), "Not yet supported");
        let start = line_set[0];
        let end = line_set[1];
        let (start, end, filled) = if start <= end {
            (start, end, true)
        } else {
            (end, start, false)
        };

        let prefix_length = usize::from(start - 1);
        let mut iter = vec![false; prefix_length];

        let stem_length = usize::from(start.abs_diff(end));
        for i in 0..=stem_length {
            if i == 0 || i == stem_length {
                iter.push(true);
            } else {
                iter.push(filled);
            }
        }

        let chunks = iter.chunks_exact(4);
        let tip = chunks.remainder().to_vec();
        let mut column: Vec<[bool; 4]> = chunks
            .into_iter()
            .map(|chunk| [chunk[0], chunk[1], chunk[2], chunk[3]])
            .collect();
        if !tip.is_empty() {
            column.push([
                tip[0],
                tip.get(1).copied().unwrap_or_default(),
                tip.get(2).copied().unwrap_or_default(),
                tip.get(3).copied().unwrap_or_default(),
            ]);
        }

        column
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn as_dot_pairs(raw: char) -> [[bool; 2]; 4] {
        let bits = u32::from(raw) - 0x2800;
        [
            [bits & 2_u32.pow(0) != 0, bits & 2_u32.pow(3) != 0],
            [bits & 2_u32.pow(1) != 0, bits & 2_u32.pow(4) != 0],
            [bits & 2_u32.pow(2) != 0, bits & 2_u32.pow(5) != 0],
            [bits & 2_u32.pow(6) != 0, bits & 2_u32.pow(7) != 0],
        ]
    }

    #[test]
    fn transpose_n3_4_2() {
        //  4  -- -- -- -*
        //  3  -- -- -- **
        //  2  -- -- -* **
        //  1  -- -- ** **
        //
        //  0  ** ** ** **
        // -1  ** *- -- --
        // -2  ** -- -- --
        // -3  *- -- -- --
        #[rustfmt::skip]
        let input = vec![
            [
                vec![[ true,  true,  true,  true], [false, false, false, false]], // -3
                vec![[false,  true,  true,  true], [false, false, false, false]], // -2
            ],
            [
                vec![[false, false,  true,  true], [false, false, false, false]], // -1
                vec![[false, false, false,  true], [false, false, false, false]], //  0
            ],
            [
                vec![[false, false, false,  true], [ true, false, false, false]], //  1
                vec![[false, false, false,  true], [ true,  true, false, false]], //  2
            ],
            [
                vec![[false, false, false,  true], [ true,  true,  true, false]], //  3
                vec![[false, false, false,  true], [ true,  true,  true,  true]], //  4
            ],
        ];

        #[rustfmt::skip]
        let expected = vec![
            vec![
                [
                    [false, false],
                    [false, false],
                    [false, false],
                    [false, false],
                ],
                [
                    [false, false],
                    [false, false],
                    [false, false],
                    [false, false],
                ],
                [
                    [false, false],
                    [false, false],
                    [false,  true],
                    [ true,  true],
                ],
                [
                    [false,  true],
                    [ true,  true],
                    [ true,  true],
                    [ true,  true],
                ],
            ],
            vec![
                [
                    [ true,  true],
                    [ true,  true],
                    [ true,  true],
                    [ true, false],
                ],
                [
                    [ true,  true],
                    [ true, false],
                    [false, false],
                    [false, false],
                ],
                [
                    [ true,  true],
                    [false, false],
                    [false, false],
                    [false, false],
                ],
                [
                    [ true,  true],
                    [false, false],
                    [false, false],
                    [false, false],
                ],
            ],
        ];

        let mut buffer = vec![];
        {
            let mut line_writer = LineWriter::new(&mut buffer);
            Columns::into_braille_rows(&mut line_writer, &input, 2).unwrap();
        }
        let dot_pairs = String::from_utf8(buffer)
            .unwrap()
            .lines()
            .map(|line| line.chars().map(as_dot_pairs).collect::<Vec<_>>())
            .collect::<Vec<_>>();

        assert_eq!(expected, dot_pairs);
    }
}
