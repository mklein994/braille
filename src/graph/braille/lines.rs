//! Turn a series of numbers into a braille graph.
//!
//! # Example
//!
//! ## Sine graph
//!
//! ```console
//! $ awk 'BEGIN { for (i = 0; i < 20; i++) { print sin(i / 3); } }' | braille 4
//! ⠀⠀⠀⠀⠀⣷⣶⣤⣀
//! ⠀⠀⠀⠀⠀⣿⣿⣿⡿⠟
//! ⠀⠀⢀⣀⣤⡟⠉⠁
//! ⣴⣿⣿⣿⣿⡇
//! ⠀⠉⠛⠻⠿⡇
//! ```
//!
//! ## Simple number sequence
//!
//! ```console
//! $ seq -4 3 | braille 7
//! ⠉⠛⠿⣿
//! ⠀⠀⠀⢸⣶⣤⣀
//! ```

use std::io::{LineWriter, Write};

use super::Brailleish;
use super::Char as BrailleChar;
use crate::graph::{BarGraphable, Graphable};
use crate::opt::{Config, GraphStyle, ValueIter};
use crate::{InputLine, InputLineSinglable};

#[derive(Debug)]
pub struct Lines {
    config: Config,
}

impl Brailleish<2> for Lines {}

impl Graphable<Option<f64>> for Lines {
    fn new(config: Config) -> Self {
        Self { config }
    }

    fn config(&self) -> &Config {
        &self.config
    }

    fn print_graph<W: Write>(
        &self,
        input_lines: ValueIter<Option<f64>>,
        mut writer: LineWriter<W>,
    ) -> anyhow::Result<()> {
        let mut input_lines = input_lines.into_iter();
        let minimum = <Self as Graphable<Option<f64>, Config>>::minimum(self);
        let maximum = <Self as Graphable<Option<f64>, Config>>::maximum(self);

        let min = 1; // reserve an empty line for null values
        let max = self.width() * 2; // braille characters are 2 dots wide
        let slope = f64::from(max - min) / (maximum - minimum);
        let scale = |value: f64| {
            assert!(
                value >= minimum && value <= maximum,
                "value out of bounds: {value} [{minimum}, {maximum}]"
            );
            min + (slope * (value - minimum)).round() as u16
        };

        // Clamp where 0 would fit to be inside the output range
        let zero = if minimum > 0. {
            min
        } else if maximum < 0. {
            max
        } else {
            scale(0.)
        };

        let style = <Self as Graphable<Option<f64>, Config>>::style(self);

        // Each braille character is 4 dots tall
        let mut buffer = [vec![], vec![], vec![], vec![]];
        let mut has_more_lines = true;
        while has_more_lines {
            for buffer_line in &mut buffer {
                let input_line = input_lines.next();
                if input_line.is_none() {
                    has_more_lines = false;
                }

                if let Some(new_line) = input_line
                    .transpose()?
                    .and_then(InputLine::into_inner)
                    .map(scale)
                    .map(|value| Self::into_dot_groups(value, zero, style))
                {
                    *buffer_line = new_line;
                }
            }

            if has_more_lines || buffer.iter().any(|x| !x.is_empty()) {
                let transposed = Self::transpose_row(&buffer);
                let braille_line = transposed
                    .into_iter()
                    .map(|x| BrailleChar::new(x).as_char())
                    .collect::<String>();
                writeln!(writer, "{braille_line}")?;
            }

            buffer.fill(vec![]);
        }

        Ok(())
    }
}

impl BarGraphable<Option<f64>> for Lines {}

impl Lines {
    #[must_use]
    pub fn into_dot_array_pairs<const N: usize>(
        line_set: [u16; N],
        style: GraphStyle,
    ) -> Vec<[bool; 2]> {
        assert_eq!(2, line_set.len(), "Not yet supported");
        let start = line_set[0];
        let end = line_set[1];

        let filled = match (start, end, style) {
            (_, _, GraphStyle::Line) => false,
            (_, _, GraphStyle::Filled) => true,
            (start, end, GraphStyle::Auto) => start <= end,
        };

        let (start, end) = if start <= end {
            (start, end)
        } else {
            (end, start)
        };

        let prefix_length = usize::from(start - 1);
        let mut iter = vec![false; prefix_length];

        let stem_length = usize::from(start.abs_diff(end));
        for i in 0..=stem_length {
            if i == 0 || i == stem_length {
                iter.push(true);
            } else {
                iter.push(filled);
            }
        }

        let chunks = iter.chunks_exact(2);
        let mut tip = chunks.remainder().to_vec();
        let mut row: Vec<[bool; 2]> = chunks
            .into_iter()
            .map(|chunk| chunk.try_into().unwrap())
            .collect();
        if !tip.is_empty() {
            tip.resize(2, false);
            row.push(tip.try_into().unwrap());
        }
        row
    }

    /// Turn a list of braille dot pairs into a list of braille dot blocks
    #[must_use]
    pub fn transpose_row(input_row: &[Vec<[bool; 2]>; 4]) -> Vec<[[bool; 2]; 4]> {
        let longest = input_row.iter().map(Vec::len).max().unwrap();

        let mut output_row = vec![];
        for column in 0..longest {
            let mut braille_character = [[false, false]; 4];

            for (row_index, row) in input_row.iter().enumerate() {
                if let Some(row_column) = row.get(column) {
                    braille_character[row_index] = *row_column;
                }
            }

            if column < longest - 1 || braille_character.into_iter().flatten().any(|x| x) {
                output_row.push(braille_character);
            }
        }

        output_row
    }
}

impl<const N: usize> Graphable<[Option<f64>; N]> for Lines {
    fn new(config: Config) -> Self {
        Self { config }
    }

    fn config(&self) -> &Config {
        &self.config
    }

    fn print_graph<W: Write>(
        &self,
        input_lines: ValueIter<[Option<f64>; N]>,
        mut writer: LineWriter<W>,
    ) -> anyhow::Result<()> {
        let mut input_lines = input_lines.into_iter();
        let minimum = <Self as Graphable<Option<f64>, Config>>::minimum(self);
        let maximum = <Self as Graphable<Option<f64>, Config>>::maximum(self);

        let min = 1; // reserve an empty line for null values
        let max = self.width() * 2; // braille characters are 2 dots wide
        let slope = f64::from(max - min) / (maximum - minimum);
        let scale = |value: f64| {
            assert!(
                value >= minimum && value <= maximum,
                "value out of bounds: {value} [{minimum}, {maximum}]"
            );
            min + (slope * (value - minimum)).round() as u16
        };

        // Each braille character is 4 dots tall
        let mut buffer = [vec![], vec![], vec![], vec![]];
        let mut has_more_lines = true;
        while has_more_lines {
            for buffer_line in &mut buffer {
                let input_line = input_lines.next();
                if input_line.is_none() {
                    has_more_lines = false;
                }

                if let Some(new_line) = input_line
                    .transpose()?
                    .and_then(|x| {
                        if x.as_single_iter().all(Option::is_none) {
                            None
                        } else {
                            let line = x.into_iter().map(|x| scale(x.unwrap())).collect::<Vec<_>>();
                            Some(<[_; N]>::try_from(line).unwrap())
                        }
                    })
                    .map(|x| {
                        Self::into_dot_array_pairs(
                            x,
                            <Self as Graphable<Option<f64>, Config>>::style(self),
                        )
                    })
                {
                    *buffer_line = new_line;
                }
            }

            if has_more_lines || buffer.iter().any(|x| !x.is_empty()) {
                let transposed = Self::transpose_row(&buffer);
                let braille_line = transposed
                    .into_iter()
                    .map(|x| BrailleChar::new(x).as_char())
                    .collect::<String>();
                writeln!(writer, "{braille_line}")?;
            }

            buffer.fill(vec![]);
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_into_dot_pairs() {
        assert_eq!(
            vec![[false, false], [false, false], [true, false]],
            Lines::into_dot_groups(5, 5, GraphStyle::default())
        );
    }

    #[test]
    fn test_transpose_row_line_1() {
        #[rustfmt::skip]
        let input = [
            vec![[ true,  true], [ true, true], [true, false]],
            vec![[false,  true], [ true, true], [true, false]],
            vec![[false, false], [ true, true], [true, false]],
            vec![[false, false], [false, true], [true, false]],
        ];

        #[rustfmt::skip]
        let expected = vec![
            [
                [ true,  true],
                [false,  true],
                [false, false],
                [false, false],
            ],
            [
                [ true,  true],
                [ true,  true],
                [ true,  true],
                [false,  true],
            ],
            [
                [ true, false],
                [ true, false],
                [ true, false],
                [ true, false],
            ],
        ];
        let actual = Lines::transpose_row(&input);
        for (ex, act) in expected.iter().zip(actual.iter()) {
            eprintln!("{ex:?}");
            eprintln!("{act:?}");
            eprintln!();
        }
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_transpose_row_line_2() {
        #[rustfmt::skip]
        let input = [
            vec![[false, false], [false, false], [true, false]               ],
            vec![[false, false], [false, false], [true,  true]               ],
            vec![[false, false], [false, false], [true,  true], [true, false]],
            vec![[false, false], [false, false], [true,  true], [true,  true]],
        ];

        #[rustfmt::skip]
        let expected = vec![
            [
                [false, false],
                [false, false],
                [false, false],
                [false, false],
            ],
            [
                [false, false],
                [false, false],
                [false, false],
                [false, false],
            ],
            [
                [ true, false],
                [ true,  true],
                [ true,  true],
                [ true,  true]
            ],
            [
                [false, false],
                [false, false],
                [ true, false],
                [ true,  true],
            ],
        ];

        assert_eq!(expected, Lines::transpose_row(&input));
    }
}
