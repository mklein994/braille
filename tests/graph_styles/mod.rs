use braille::{GraphKind, GraphStyle};

use crate::tt;

fn get_values() -> Vec<f64> {
    use std::f64::consts::PI;
    let mut values = vec![];
    let mut i = -8. * PI;
    while i < 8. * PI {
        values.push(i);
        i += 1.;
    }
    values
}

fn curve1() -> Vec<f64> {
    get_values().into_iter().map(|x| f64::cos(x / 5.)).collect()
}

fn curve2() -> Vec<f64> {
    get_values().into_iter().map(|x| f64::sin(x / 4.)).collect()
}

fn get_both_curves() -> Vec<(f64, f64)> {
    std::iter::zip(curve1(), curve2()).collect()
}

fn get_input(kind: GraphKind, style: GraphStyle, per: u8) -> String {
    use std::fmt::Write;

    assert!(
        matches!(
            (kind, style, per),
            (
                GraphKind::Bars | GraphKind::BrailleBars | GraphKind::BrailleColumns,
                _,
                1 | 2
            ) | (GraphKind::Columns, GraphStyle::Filled, 1)
        ),
        "invalid flag combo"
    );

    let kind_flag = match kind {
        GraphKind::Bars => "-B",
        GraphKind::Columns => "-C",
        GraphKind::BrailleBars => "-b",
        GraphKind::BrailleColumns => "-c",
    };

    let style_flag = match style {
        GraphStyle::Auto => "-sa",
        GraphStyle::Line => "-sl",
        GraphStyle::Filled => "-sf",
    };

    let per_flag = match per {
        1 => "",
        2 => "-p2",
        _ => unreachable!(),
    };

    let modeline = ["braille", kind_flag, style_flag, per_flag, "10"].join(" ");

    let values = if per == 1 {
        curve1()
            .into_iter()
            .map(|x| x.to_string())
            .collect::<Vec<_>>()
    } else {
        get_both_curves()
            .into_iter()
            .map(|(left, right)| format!("{left}\t{right}"))
            .collect::<Vec<_>>()
    };

    let input = values
        .into_iter()
        .fold(format!("{modeline}\n"), |mut input, line| {
            let _ = writeln!(input, "{line}");
            input
        });

    input.trim_end().to_string()
}

#[test]
fn check_against_jq() {
    use std::process::{Command, Stdio};

    let rust_values: Vec<(f64, f64)> = get_both_curves();
    let jq_output = String::from_utf8(
        Command::new("jq")
            .arg("-nrf")
            .arg(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/curve-2.jq"))
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .output()
            .unwrap()
            .stdout,
    )
    .unwrap();

    let jq_values: Vec<(f64, f64)> = jq_output
        .lines()
        .map(|line| -> Result<_, std::num::ParseFloatError> {
            let (left, right) = line.split_once('\t').unwrap();
            Ok((left.parse()?, right.parse()?))
        })
        .collect::<Result<_, _>>()
        .unwrap();

    assert_eq!(rust_values, jq_values);
}

macro_rules! t {
    (#style_test $name:ident, $kind:expr, $per:literal) => {{
        tt!(#concat!(stringify!($name), "_", stringify!($per), "_auto"), get_input($kind, Auto, $per), ["-m"]);
        tt!(#concat!(stringify!($name), "_", stringify!($per), "_filled"), get_input($kind, Filled, $per), ["-m"]);
        tt!(#concat!(stringify!($name), "_", stringify!($per), "_line"), get_input($kind, Line, $per), ["-m"]);
    }};

    ($name:ident, $kind:expr) => {
        #[test]
        fn $name() {
            use GraphStyle::*;

            t!(#style_test $name, $kind, 1);
            t!(#style_test $name, $kind, 2);
        }
    };

    ($name:ident, $kind:expr, $style:expr, $per:expr) => {
        tt!($name, get_input($kind, $style, $per), ["-m"]);
    };
}

t!(test_columns, GraphKind::Columns, GraphStyle::default(), 1);

t!(test_bars, GraphKind::Bars);
t!(test_braille_columns, GraphKind::BrailleColumns);
t!(test_braille_bars, GraphKind::BrailleBars);
