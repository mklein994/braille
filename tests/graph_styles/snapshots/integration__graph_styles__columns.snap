---
source: tests/graph_styles/mod.rs
expression: stdout
---
                       ▄▇█▇▅▁                      
                     ▃███████▄                     
                    ▄█████████▆                    
▄                  ▅███████████▇                  ▂
█▄                ▆██████████████                ▂█
██▅              ▆████████████████              ▃██
███▅            ▆██████████████████▁           ▃███
████▆          ▇████████████████████▂         ▄████
█████▇▁      ▂███████████████████████▄       ▅█████
███████▅▂▁▁▂▆█████████████████████████▇▃▁▁▂▄███████

